egs/wsj/s5/steps/nnet3/get_egs_discriminative.sh

```diff
@@ -0,0 +1,461 @@
+  # files is the product of 'nj' by 'num_archives_intermediate', which might be
+  # quite large.
+  $cmd --max-jobs-run $max_jobs_run JOB=1:$nj $dir/log/get_egs.JOB.log \
+    discriminative-get-supervision $supervision_all_opts \
+    "scp:utils/filter_scp.pl $sdata/JOB/utt2spk $dir/ali.scp |" \
+    "ark:gunzip -c $denlatdir/lat.JOB.gz |" ark:- \| \
+    nnet3-discriminative-get-egs $ivector_opt $egs_opts \
+    $dir/final.mdl "$feats" ark,s,cs:- ark:- \| \
+    nnet3-discriminative-copy-egs --random=true --srand=JOB ark:- $degs_list || exit 1;
+fi
+
+if [ $stage -le 6 ]; then
+  echo "$0: recombining and shuffling order of archives on disk"
+  # combine all the "degs_orig.*.JOB.scp" (over the $nj splits of the data) and
+  # shuffle the order, writing to the degs.JOB.ark
+
+  # the input is a concatenation over the input jobs.
+  degs_list=
+  for n in $(seq $nj); do
+    degs_list="$degs_list $dir/degs_orig.$n.JOB.ark"
+  done
+  
+  if [ $archives_multiple == 1 ]; then # normal case.
+    $cmd --max-jobs-run $max_shuffle_jobs_run --mem 8G JOB=1:$num_archives_intermediate $dir/log/shuffle.JOB.log \
+      nnet3-discriminative-shuffle-egs --srand=JOB "ark:cat $degs_list|" ark:$dir/degs.JOB.ark  || exit 1;
+  else
+    # we need to shuffle the 'intermediate archives' and then split into the
+    # final archives.  we create soft links to manage this splitting, because
+    # otherwise managing the output names is quite difficult (and we don't want
+    # to submit separate queue jobs for each intermediate archive, because then
+    # the --max-jobs-run option is hard to enforce).
+    output_archives="$(for y in $(seq $archives_multiple); do echo ark:$dir/degs.JOB.$y.ark; done)"
+    for x in $(seq $num_archives_intermediate); do
+      for y in $(seq $archives_multiple); do
+        archive_index=$[($x-1)*$archives_multiple+$y]
+        # degs.intermediate_archive.{1,2,...}.ark will point to degs.archive.ark
+        ln -sf degs.$archive_index.ark $dir/degs.$x.$y.ark || exit 1
+      done
+    done
+    $cmd --max-jobs-run $max_shuffle_jobs_run --mem 8G JOB=1:$num_archives_intermediate $dir/log/shuffle.JOB.log \
+      nnet3-discriminative-shuffle-egs --srand=JOB "ark:cat $degs_list|" ark:- \| \
+      nnet3-discriminative-copy-egs ark:- $output_archives || exit 1;
+  fi
+fi
+  
+if [ $stage -le 7 ]; then
+  echo "$0: removing temporary archives"
+  (
+    cd $dir
+    for f in $(ls -l . | grep 'degs_orig' | awk '{ X=NF-1; Y=NF-2; if ($X == "->")  print $Y, $NF; }'); do rm $f; done
+  )
+  if [ $archives_multiple -gt 1 ]; then
+    # there are some extra soft links that we should delete.
+    for f in $dir/degs.*.*.ark; do rm $f; done
+  fi
+  echo "$0: removing temporary lattices"
+  rm $dir/lat.*
+  echo "$0: removing temporary alignments and transforms"
+  # Ignore errors below because trans.* might not exist.
+  rm -f $dir/{ali,trans}.{ark,scp} 2>/dev/null || true
```

Comment:
please never add -f to 'rm' statements.  the reason we do 2>/dev/null is to suppress warnings.
if someone does chmod -w [file], then we have to assume they did it for a reason, and respect that.  hence no -f.  check elsewhere.